libs/client/src/client/PezzoOpenAI.ts

```diff
@@ -101,47 +102,85 @@ class Completions {
     const baseReport = {
       metadata: merge(baseMetadata, pezzoPrompt?.metadata),
       properties: pezzoOptions?.properties,
+      cacheEnabled: false,
+      cacheHit: null,
       request: {
         timestamp: requestTimestamp,
         body: requestBody,
       },
     };
 
-    try {
-      response = await this.openai.chat.completions.create(
-        {
-          ...(requestBody as OpenAI.Chat.CompletionCreateParamsNonStreaming),
-        },
-        "variables" in optionsOrPezzoProps
-          ? undefined
-          : (optionsOrPezzoProps as Parameters<
-              OpenAI["chat"]["completions"]["create"]
-            >[1])
-      );
-
-      reportPayload = {
-        ...baseReport,
-        response: {
-          timestamp: new Date().toISOString(),
-          body: response,
-          status: 200,
-        },
-      };
-    } catch (err) {
-      error = err;
-
-      reportPayload = {
-        ...baseReport,
-        response: {
-          timestamp: new Date().toISOString(),
-          body: err.error,
-          status: err.status,
-        },
-      };
+    if (pezzoOptions.cache) {
+      baseReport.cacheEnabled = true;
+
+      const cachedRequest = await this.pezzo.fetchCachedRequest(requestBody);
+
+      if (cachedRequest.hit === true) {
+        baseReport.cacheHit = true;
+        response = cachedRequest.data;
+
+        reportPayload = {
+          ...baseReport,
+          response: {
+            timestamp: new Date().toISOString(),
+            body: response,
+            status: 200,
+          },
+        };
+      } else {
+        baseReport.cacheHit = false;
+      }
     }
 
+    if (!pezzoOptions.cache || (pezzoOptions.cache && !baseReport.cacheHit)) {
+      try {
+        response = await this.openai.chat.completions.create(
+          {
+            ...(requestBody as OpenAI.Chat.CompletionCreateParamsNonStreaming),
+          },
+          "variables" in optionsOrPezzoProps
+            ? undefined
+            : (optionsOrPezzoProps as Parameters<
+                OpenAI["chat"]["completions"]["create"]
+              >[1])
+        );
+
+        reportPayload = {
+          ...baseReport,
+          response: {
+            timestamp: new Date().toISOString(),
+            body: response,
+            status: 200,
+          },
+        };
+      } catch (err) {
+        error = err;
+
+        reportPayload = {
+          ...baseReport,
+          response: {
+            timestamp: new Date().toISOString(),
+            body: err.error,
+            status: err.status,
+          },
+        };
+      }
+    }
+
+    const promises = [];
+
+    if (
```

Comment:
I personally prefer more declarative approach, `promises.push` specifically. What do you say about this?
```suggestion
    const shouldCacheRequest =    pezzoOptions.cache &&
      reportPayload.cacheHit === false &&
      reportPayload.response.status === 200
      
      const reportRequest = this.pezzo.reportPromptExecution(reportPayload);
      await Promise.all(shouldCache ? [this.pezzo.cacheRequest(requestBody, response), reportRequest] : [reportRequest])
```